File: src/sg/nphau/leetcode/easy/ValidParentheses.java
/**
 * Created by nphau on 09/12/2021, 00:30
 * Copyright (c) 2021 . All rights reserved.
 * Last modified 09/12/2021, 15:54
 */
package sg.nphau.leetcode.easy;

import java.util.Stack;

/**
 * https://leetcode.com/problems/valid-parentheses
 */
public class ValidParentheses {
    public boolean isValid(String s) {

        if (s == null || s.length() == 0) {
            return true;
        }

        if (s.length() % 2 == 1) {
            return false;
        }

        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            if (c == '(')
                stack.push(')');
            else if (c == '{')
                stack.push('}');
            else if (c == '[')
                stack.push(']');
            else if (stack.isEmpty() || stack.pop() != c)
                return false;
        }
        return stack.isEmpty();

    }
}

File: src/com/nphausg/leetcode/easy/ValidParentheses.kt
package com.nphausg.leetcode.easy

import java.util.ArrayDeque

/**
 * <a href="https://leetcode.com/problems/valid-parentheses">20. Valid Parentheses</a>
 */
class ValidParenthesesKt {

    fun isValid(s: String): Boolean {
        if (s.length % 2 == 1) {
            return false
        }

        val stack = ArrayDeque<Char>()
        for (char in s) {
            when (char) {
                '(' -> stack.addLast(')')
                '{' -> stack.addLast('}')
                '[' -> stack.addLast(']')
                else -> if (stack.isEmpty() || stack.removeLast() != char) {
                    return false
                }
            }
        }

        return stack.isEmpty()
    }

    fun isValidMapBased(s: String): Boolean {
        if (s.length % 2 == 1) {
            return false
        }

        val pairs = mapOf(')' to '(', '}' to '{', ']' to '[')
        val stack = ArrayDeque<Char>()

        for (char in s) {
            if (char in pairs.values) {
                stack.addLast(char)
            } else {
                if (stack.isEmpty() || stack.removeLast() != pairs[char]) {
                    return false
                }
            }
        }

        return stack.isEmpty()
    }
}
